content/blog/2025-04-16-unified-gc-theory.md

```diff
@@ -0,0 +1,110 @@
+design space for GCs and better select the best GC strategy for a particular application. 
+
+# Contributions
+
+## Qualitative analysis
+By presenting tracing and RC as duals, the authors introduce a novel mental model for approaching garbage collectors. 
+Specifically, tracing operates on live objects (“matter”), while RC operates on dead objects (“anti-matter”). Concretely, tracing initializes reference counts to 0 (an underestimate of the true count), *incrementing* them during graph traversal until they reach the true count. On the other hand, RC initializes reference counts to an overestimate of the true count, *decrementing* them during graph traversal until they reach the true count (ignoring cycles). (With RC, we start with an overestimate since we count in-edges from objects that are no longer live.) 
+
+In addition, the authors formulate garbage collection as a fix-point problem, and they demonstrate that 
+tracing computes the *least* fix point, while RC computes the *greatest* fix point, with their set difference being cyclic garbage.
+
+The authors also identify several key characteristics of each algorithm:
+| | Tracing | RC |
+| --- | --- | --- |
+| Starting point | Roots | Anti-roots |
+| Graph traversal | Forward from roots | Forward from anti-roots | 
+| Objects traversed | Live | Dead |
+
+With this in mind, the authors leave the reader with 3 considerations to keep in mind when designing a new GC algorithm:
+1. **Partition**: Should memory be divided into different regions which are each subjected to (possibly different) strategies?
+2. **Traversal**: For each partition, should tracing or RC be used?
+3. **Trade-offs**: For each partition, decide how to handle space-time trade-offs
+
+## Hybrid collectors
+The authors demonstrate that in practice, various GC strategies lie on a continuum between tracing and RC. For example, 
+
+For collectors where there is a *unified heap* (i.e. a single heap in which all data resides), we have:
+- **Deferred RC (DRC)**: References from the stack to the heap are traced, while references within the heap are reference-counted
+- **Partial tracing**: Converse of DRC, i.e. reference-count roots and trace the heap
+
+For generational GCs, where the heap is split into a nursery and a mature space, we have variations where:
+- both the nursery and mature space are traced (standard generational GC)
+- the nursery is RCed and the mature space is traced 
+- the nursery is traced and the mature space is reference-counted (Ulterior Reference Counting)
+
+## Cost model
+The authors also present a formal cost model for comparing the performance characteristics of different collectors. Specifically, they introduce:
+- $\kappa$, the time overhead for a single garbage collection
```

Comment:
To make the math show up, you'll need to add `latex = true` to your front matter (see the [README](https://github.com/sampsyo/cs6120/blob/2025sp/README.md)).